core/environment/environment.go

```diff
@@ -208,4 +208,15 @@ func (env *Environment) GetCurrentRunNumber() (rn uint32) {
 		return 0
 	}
 	return env.currentRunNumber
+}
+
+// setState will move environment to a given state from current state.
+// The call does not trigger any callbacks, if defined.
+func (env *Environment) setState(state string) {
+	if env == nil {
+		return
+	}
+	env.Mu.RLock()
```

Comment:
Since we're writing to the `Sm`, I suspect we should use `Lock` instead of `RLock` here.